core/trino-main/src/main/java/io/trino/operator/ScanFilterAndProjectOperator.java

```diff
@@ -157,6 +157,14 @@ public long getDynamicFilterSplitsProcessed()
         return dynamicFilterSplitsProcessed;
     }
 
+    @Override
+    public synchronized void cancel()
```

Comment:
What are you synchronizing here for?

add relevant `@GuardedBy`
and add synchronization in close() as well

however, i actually think `volatile` would be better, as it would express the non-blocking intent you have here

Reply:
That was intentional. Operator is not synchronized other than in this two places so I don't won't open critical section for every access of `cursor` field.

For that reason I didn't add `@GuardedBy`.